dotCMS/src/main/java/com/dotcms/api/content/VanityUrlAPIImpl.java

```diff
@@ -0,0 +1,197 @@
+			contentResults.stream().forEach((Contentlet con) -> {
+				results.add(getVanityUrlFromContentlet(con));
+			});
+		} catch(DotDataException | DotSecurityException e){
+			Logger.error(this,"Error searching vanity URLs",e);
+		}
+		return results.build();
+	}
+
+	@Override
+	public VanityUrl getWorkingVanityUrl(final String uri, final Host host, final long languageId, final User user) {
+		return getVanityUrlByURI(uri,host,languageId,user,false);
+	}
+
+	@Override
+	public VanityUrl getLiveVanityUrl(final String uri, final Host host, final long languageId, final User user) {
+		return getVanityUrlByURI(uri,host,languageId,user,true);
+	}
+
+	protected VanityUrl getVanityUrlByURI(final String uri, final Host host, final long languageId, final User user, final boolean live) {
+		VanityUrl result = vanityURLCache.get(VanityUrlUtil.sanitizeKey(host != null && InodeUtils.isSet(host.getInode())?host.getIdentifier()+"|"+VanityUrlUtil.fixURI(uri):VanityUrlUtil.fixURI(uri),languageId));
+		if(result == null || !InodeUtils.isSet(result.getInode())){
+			List<VanityUrl> results = new ArrayList<>();
+			String hostCondition = (host != null?" +conHost:"+host.getIdentifier():" +conHost:"+Host.SYSTEM_HOST);
+			try {
+                List<Contentlet> contentResults = contentletAPI.search("+baseType:" + BaseContentType.VANITY_URL.getType() + " +languageId:" + languageId + hostCondition + " +vanityUrl:" + VanityUrlUtil.fixURI(uri) + (live ? " +live:true" : " +working:true"), 0, 0, StringPool.BLANK, user, false);
+                contentResults.stream().forEach((Contentlet con) -> {
+                    VanityUrl vanityUrl = getVanityUrlFromContentlet(con);
+                    try {
+                        if (con.isLive()) {
+                            addToVanityURLCache(vanityUrl);
+                        } else {
+                            VanityUrlServices.invalidateVanityUrl(vanityUrl);
+                        }
+                    } catch (DotDataException | DotSecurityException e) {
+                        Logger.error(this, "Error processing Vanity Url - contentlet Id:" + con.getIdentifier(), e);
+                    }
+                    results.add(vanityUrl);
+                });
+
+                List<ContentType> vanityUrlContentTypes = APILocator.getContentTypeAPI(user).findByType(BaseContentType.VANITY_URL);
+
+				if(results.size() == 0 && vanityUrlContentTypes.size() > 0 && ((VanityUrlContentType)vanityUrlContentTypes.get(0)).fallback()){
+					contentResults = contentletAPI.search("+baseType:" + BaseContentType.VANITY_URL.getType() + " +languageId:" + APILocator.getLanguageAPI().getDefaultLanguage().getId() + hostCondition + " +vanityUrl:" + VanityUrlUtil.fixURI(uri) + (live ? " +live:true" : " +working:true"), 0, 0, StringPool.BLANK, user, false);
+                    contentResults.stream().forEach((Contentlet con) ->{
```

Comment:
I think is better to use the old for loops instead the stream.foreach. (Performance and readability)